File: java-app/src/main/java/com/mercadolibre/java_app/services/shipping/ShippingOptionServices.java
package com.mercadolibre.java_app.services.shipping;

import com.mercadolibre.java_app.models.shipping.ShippingOption;

import java.util.List;

public final class ShippingOptionServices implements ShippingOptionsInteractor {

    @Override
    public ShippingOption retrieveBestShippingOption() {
        List<ShippingOption> shippingOptions = ShippingOptionsDataSource.retrieveShippingOptions();
        ShippingOption bestShippingOption = null;

        for (ShippingOption shippingOption : shippingOptions) {
            bestShippingOption = shippingOption.getBestShippingOption(bestShippingOption);
        }

        return bestShippingOption;
    }
}

File: kt-app/src/main/java/com/mercadolibre/melikotlinmeetup/services/shipping/ShippingOptionServices.kt
package com.mercadolibre.melikotlinmeetup.services.shipping

import com.mercadolibre.melikotlinmeetup.commons.ShippingOption

class ShippingOptionServices: ShippingOptionsInteractor {
    override fun retrieveBestShippingOption(): ShippingOption? {
        val shippingOptions = retrieveShippingOptions()
        var bestShippingOption: ShippingOption? = null

        for (shippingOption in shippingOptions) {
            bestShippingOption = shippingOption.getBestShippingOption(bestShippingOption)
        }

        return bestShippingOption
    }
}